### ISSUE TITLE
Slow training speed — fixed by limiting thread count in DataLoader workers

### ISSUE THREAD

**Issue by m-mgk** [eduardoAcunaEspinoza/Hy2DL#48]
Hi, I wanted to report a performance issue I ran into, along with the fix that worked for me. I used the dev branch and maybe I just faced this problem but I tried it on two systems and on both, I faced this issue.

### Setup

I'm using the provided config file for the hourly resolution of CAMELS-DE, with seed 100 (config file is there when you download the hourly_camels_de dataset in the ensemble_runs folder) . The only difference from the default is that I'm training on 114 catchments instead of the full set and use path_data (not using the path_dataset_training which takes the pre-processed zarr formats)

I ran this on two systems:

- **HPC node with an NVIDIA H100** (TU Dresden Capella)
- **Local machine with an NVIDIA RTX 5070** (24 CPU cores)

### The problem

Training was much slower than expected. On the HPC, each epoch was taking around **44 minutes**, even though the GPU usage stayed very low (around 2-3 GB) and the GPU seemed mostly idle.

Increasing `num_workers` didn't help — and on the HPC, requesting more CPU cores (going from 8 up to 28) didn't improve the epoch time at all, even though all the CPUs showed high usage. When I looked at the processes with `htop`, each DataLoader worker was using more than 140% CPU, which suggested the workers were each spawning multiple internal threads and competing with each other for CPU. This seemed to starve the main process so it couldn't keep the GPU busy.

### The fix

Setting these environment variables in the very first cell of the notebook, **before importing anything else**, fixed it:

```python
import os
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"
```

After this, each worker uses a single thread, the CPU contention goes away, and the GPU gets fed properly.

One important detail: this only works if the cell runs **before** numpy/torch are imported, in a fresh kernel. 

### Results (114 basins)

| System | num_workers | Before | After |
|---|---|---|---|
| HPC (H100) | 16 | ~44 min/epoch | ~8.5 min/epoch |
| Local (RTX 5070) | 16 | ~33 min/epoch | ~21 min/epoch |

On the HPC, I get the same ~8.5 min/epoch with 16 cores as I do with higher core counts, so the threading fix is what matters, not the number of cores.

The training results are unchanged — I checked and the loss values are identical with and without the fix, so this is purely a speed improvement, not anything that affects the model.

### Environment (local machine with RTX 5070)

- Python 3.12.3
- torch 2.11.0+cu128 (CUDA 12.8, cuDNN 9.19)
- numpy 2.4.6, pandas 3.0.3, xarray 2026.2.0, dask 2026.3.0, zarr 2.18.7, scipy 1.17.1
- GPU: NVIDIA RTX 5070
- OS: Linux 6.17 (Ubuntu)
- Hy2DL 2.0.0 (editable, dev_scaling branch)

on my HPC, 

**HPC (Capella):**
- Python 3.12.3
- torch 2.6.0+cu124 (CUDA 12.4)
- numpy 2.4.4, pandas 3.0.3, xarray 2026.2.0, dask 2026.3.0, zarr 2.18.7
- GPU: NVIDIA H100 (Hopper, sm_90)
- Hy2DL 2.0.0 (editable, dev_scaling branch)

Note: 
I was this comment on the main branch, so may this is fixed but I updated my dev branch yesterday (June 08, 2026) and tested it and still faced the issue.

Optimized Memory & Scalability: Added Zarr support, enabling lazy loading to handle massive datasets. Additionally, a critical memory explosion issue was resolved when utilizing multiple workers.

Regards,
Musa

**Comment by eduardoAcunaEspinoza** [comment]
Hi Musa,

Thank you for reporting the problem, and for making such a detailed issue report.  To make the solution a bit more permanent I think we can include the lines you proposed in the **__init__** of the librabry, and in a function that is called when we create the dataloaders with multiple workers.  For that, there are only two changes that need to be made:

Add on **src/hy2dl/__init__.py/**

>     import os
>     os.environ.setdefault("OMP_NUM_THREADS", "1")
>     os.environ.setdefault("MKL_NUM_THREADS", "1")
>     os.environ.setdefault("OPENBLAS_NUM_THREADS", "1")
>     os.environ.setdefault("NUMEXPR_NUM_THREADS", "1")

Add on **src/hy2dl/datasetzoo/basedataset.py/**

>     @staticmethod
>     def dask_worker_init_fn(worker_id):
>         """Initialization function for Dask workers.
> 
>         Note: This function is called inside each PyTorch worker
> 
>         """
>         os.environ.setdefault("OMP_NUM_THREADS", "1")
>         os.environ.setdefault("MKL_NUM_THREADS", "1")
>         os.environ.setdefault("OPENBLAS_NUM_THREADS", "1")
>         os.environ.setdefault("NUMEXPR_NUM_THREADS", "1")
> 
>         dask.config.set(scheduler="synchronous")


Can you check if making these two changes, plus removing the lines you added to the first cell of your notebook, solves the issue?


Let me know!

Kind regards,
Eduardo

**Comment by AlexDo1** [comment]
Hi Musa,

First of all thank you for reporting in so much detail about this issue. I talked to Eduardo and I will just jump in here and leave a comment.  
I think the changes Eduardo proposed can in some cases solve the issue, but only if hy2dl is really the first library that is imported in any script or notebook that someone uses (e.g. if numpy is imported first, it uses the wrong / unset environment variables). That's something that we cannot rely on.

What you encountered is a know problem that is solved by a library, which is maintained by the developers of scipy and scikit-learn: [threadpoolctl](https://github.com/joblib/threadpoolctl)
This would add another dependency but is considered best practice to solve this issue.

I think to implement this we would need to change this function:
https://github.com/eduardoAcunaEspinoza/Hy2DL/blob/8aa7c90debb5686e69b8fbf85148a61ae2d739f0/src/hy2dl/datasetzoo/basedataset.py#L1683-L1690

to 
```python
from threadpoolctl import threadpool_limits

...

@staticmethod
def dask_worker_init_fn(worker_id):
    # This reaches directly into the already-loaded C-libraries 
    # and forces them to 1 thread, bypassing the environment variables entirely.
    global _thread_limits # Keep a reference so it isn't garbage collected
    _thread_limits = threadpool_limits(limits=1)
```

It would be very helpful if you can try this in your setup.

Best wishes,
Alex

**Comment by m-mgk** [comment]
Hi, thank you  for the suggestion!

I first tried your version: `threadpool_limits(limits=1)` inside `dask_worker_init_fn`. I confirmed (by writing each worker's `threadpool_info()` to a file) that all workers were limited to 1 thread. But training time didn't improve — still ~33 min/epoch on my RTX 5070 system.

Looking at the **main process** showed why: numpy's OpenBLAS was still at 24 threads (and torch's OpenMP at 17). Since `dask_worker_init_fn` only runs in the workers, it never limits the main process — and that's where the bottleneck was.

So I limited the main process instead, with one line at module level in `basedataset.py`:

```python
from threadpoolctl import threadpool_limits
_main_thread_limits = threadpool_limits(limits=1)
```

With only this (and `dask_worker_init_fn` left at its original), training dropped to ~21 min/epoch — same as the env var approach.

I also noticed the workers were already limited to 1 thread **without** any threadpoolctl line in `dask_worker_init_fn` — I think PyTorch sets `OMP_NUM_THREADS=1` for DataLoader workers by default. So the worker-side change is redundant; the main process was the only thing that needed limiting.

**Takeaway:** the limit needs to go in the main process, not the workers.

For context, my machine has 24 cores, and numpy's bundled OpenBLAS (libscipy_openblas64) defaults to using all 24 threads in the main process — that's the oversubscription that was slowing things down. The single threadpool_limits(limits=1) call catches that plus torch's OpenMP and a separate scipy OpenBLAS.

One thing to consider for a library-wide change: limiting the main process globally also single-threads any BLAS-heavy numpy work during dataset setup/standardization. For my dataset that's negligible, but maybe it matters for larger datasets.

Regards,
Musa

**Comment by eduardoAcunaEspinoza** [comment]
Hi Musa,

Again, thank you for the detailed solution!  If possible, I would like to ask you one additional test, to see if we can restrict the problem during training (and maybe evaluation)  but leave the multi-threads option while we prepare the data. 

Can you try setting the treadpool_limit to 1 in the training loop:
**hy2dl/training/basetrainer.py**
>         def train_model(self,epoch):
>              Loop through the batches
>              running_loss = 0.0
>              training_time = time.time()
>         
>             with threadpool_limits(limits=1):     # <-----  HERE!
>                 for idx, sample in enumerate(iterator):
>                     if self.cfg.max_updates_per_epoch is not None and idx >= self.cfg.max_updates_per_epoch:
>                         break  # reach maximum iterations per epoch
>                     sample = upload_to_device(sample, self.cfg.device)  # upload tensors to device
>                     self.optimizer.optimizer.zero
>            # ... 

And in the evaluation loop:
**hy2dl/evaluation/basetester.py**

>         def evaluate_model(self, model: torch.nn.Module):
>              model.eval()
>              self._initialize_zarr()
>              self.gauge_data = {k: [] for k in self.gauge_data}
>              current_gauge = None
>              with torch.no_grad():
>                 # ....
>                 with threadpool_limits(limits=1):  # <-----  HERE!           
>                      for sample in iterator:
>                      sample = upload_to_device(sample, self.cfg.device)
>                      batch_gauge_id = sample["gauge_id"][0]



Thanks in advance!

**Comment by m-mgk** [comment]
Tested it — works perfectly, ~21 min/epoch (down from ~33) on my local RTX 5070 machine, same as the global approach or putting the variables in the first cell of notebook.

Before testing, I reset all my previous edits.

I didn't do detailed verification this time of whether the threads are actually at 1 *during* the training loop, but the fast training time strongly suggests they are — it matches exactly what I got when I limited the main process explicitly. (For what it's worth, if I check `threadpool_info()` outside the loop it shows 24 again, which makes sense since the context manager scopes the limit to just the loop.)

Thanks!
